uqcsbot/advent.py

```diff
@@ -183,296 +288,795 @@ def __init__(self, bot: UQCSBot):
+                    for member in potential_winners
+                ]
+            case "Equal":
+                weight_values = [1 for _ in potential_winners]
+
+        if allow_repeat_winners:
+            winners = choices(potential_winners, weight_values, k=number_of_winners)
+        else:
+            winners = self._random_choices_without_repition(
+                potential_winners, weight_values, number_of_winners
             )
-            raise
 
-        try:
-            members = [
-                Member.from_member_data(data, args.year, args.day)
-                for data in leaderboard["members"].values()
-            ]
-        except Exception:
-            await ctx.send("Error parsing leaderboard data.")
-            raise
+        if not winners:
+            await interaction.edit_original_response(
+                content="There was some problem choosing the winners."
+            )
+            return
+
+        self._add_winners(winners, year, prize)
+
+        distinct_winners = set(winners)
+        if len(distinct_winners) == 1:
+            (winner,) = distinct_winners
+            discord_id = winner.get_discord_userid(self.bot)
+            discord_ping = f" (<@{discord_id})" if discord_id else ""
+            await interaction.edit_original_response(
+                content=f"The results are in! Out of {len(potential_winners)} potential participants, {winner.name}{discord_ping} has recieved a prize from participating in Advent of Code: {prize}"
+            )
+            return
+
+        winners_message = ""
+        for i, winner in enumerate(distinct_winners):
+            discord_id = winner.get_discord_userid(self.bot)
+            discord_ping = f" (<@{discord_id})" if discord_id else ""
+            number_of_prizes = len(
+                [member for member in winners if member.id == winner.id]
+            )
+            prize_multiplier = f" (x{number_of_prizes})" if number_of_prizes > 1 else ""
+            winners_message += f"{winner.name}{discord_ping}{prize_multiplier}"
+            winners_message += ", " if i < len(distinct_winners) - 1 else " and "
+
+        await interaction.edit_original_response(
+            content=f"The results are in! Out of {len(potential_winners)} potential participants, {winners_message} have recieved a prize from participating in Advent of Code: {prize}"
+        )
 
-        # whether to show only one day
-        is_day = bool(args.day)
-        # whether to use global points
-        is_global = args.global_
+    @app_commands.default_permissions(manage_guild=True)
+    @advent_command_group.command(name="remove-winner")
+    @app_commands.describe(
+        id="The database entry id for the winners database that should be deleted."
+    )
+    async def remove_winner_command(self, interaction: discord.Interaction, id: int):
+        """
+        Remove an AOC winner from the database.
+        The show_ids option for previous-winners can get the id.
```

Comment:
See above comment about using discord IDs as identifiers - it removes the need for this whole system of getting, storing, and updating using our own custom IDs